v2/ical.NET/Serialization/SimpleDeserializer.cs

```diff
@@ -0,0 +1,221 @@
+
+            // param-value   = paramtext / quoted-string
+            // paramtext     = *SAFE-CHAR
+            // quoted-string = DQUOTE *QSAFE-CHAR DQUOTE
+            // QSAFE-CHAR    = WSP / %x21 / %x23-7E / NON-US-ASCII
+            // ; Any character except CONTROL and DQUOTE
+            // SAFE-CHAR     = WSP / %x21 / %x23-2B / %x2D-39 / %x3C-7E
+            //               / NON-US-ASCII
+            // ; Any character except CONTROL, DQUOTE, ";", ":", ","
+            var paramValue = $"((?<{_paramValueGroup}>[^\\x00-\\x08\\x0A-\\x1F\\x7F\";:,]*)|\"(?<{_paramValueGroup}>[^\\x00-\\x08\\x0A-\\x1F\\x7F\"]*)\")";
+
+            // param         = param-name "=" param-value *("," param-value)
+            // param-name    = iana-token / x-name
+            var paramName = $"(?<{_paramNameGroup}>{identifier})";
+            var param = $"{paramName}={paramValue}(,{paramValue})*";
+
+            // contentline   = name *(";" param ) ":" value CRLF
+            var name = $"(?<{_nameGroup}>{identifier})";
+            // value         = *VALUE-CHAR
+            var value = $"(?<{_valueGroup}>[^\\x00-\\x19]*)";
+            var contentLine = $"^{name}(;{param})*:{value}$";
+            return contentLine;
+        }
+
+        public IEnumerable<ICalendarComponent> Deserialize(TextReader reader)
+        {
+            var context = new SerializationContext();
+            var stack = new Stack<ICalendarComponent>();
+            var current = default(ICalendarComponent);
+            foreach (var contentLineString in GetContentLines(reader))
+            {
+                var contentLine = ParseContentLine(context, contentLineString);
+                if (string.Equals(contentLine.Name, "BEGIN", StringComparison.OrdinalIgnoreCase))
+                {
+                    stack.Push(current);
+                    current = _componentFactory.Build((string)contentLine.Value);
+                    SerializationUtil.OnDeserializing(current);
+                }
+                else
+                {
+                    if (current == null)
+                    {
+                        throw new SerializationException($"Expected 'BEGIN', found '{contentLine.Name}'");
+                    }
+                    if (string.Equals(contentLine.Name, "END", StringComparison.OrdinalIgnoreCase))
+                    {
+                        if (!string.Equals((string)contentLine.Value, current.Name, StringComparison.OrdinalIgnoreCase))
+                        {
+                            throw new SerializationException($"Expected 'END:{current.Name}', found 'END:{contentLine.Value}'");
+                        }
+                        SerializationUtil.OnDeserialized(current);
+                        var finished = current;
+                        current = stack.Pop();
+                        if (current == null)
+                        {
+                            yield return finished;
```

Comment:
There is an extra `null` reference lying around.  `current` is initialized to `null` when we're outside of any components, so the first `stack.Push(current);` pushes the `null` onto the stack.